### DISCUSSION TITLE
Interpolate from CSV

### DISCUSSION THREAD

**Discussion by hsheldon** [idaholab/moose#20480] [Q&A General] [answered]
Is there a way to interpolate values stored in a CSV file onto a mesh? e.g. I have a CSV containing xyz coordinates and corresponding values of various parameters at those locations, and I would like to use these to set the values of AuxKernels on my mesh.
Cheers,
Heather

**Comment by WilkAndy** [comment, marked as answer]
Hi @hsheldon , long time no see.... are your values stored at a grid.  If so, PiecewiseMultilinear might be helpful?  https://mooseframework.inl.gov/source/functions/PiecewiseMultilinear.html

**Reply by hsheldon** [reply]
Hi Andy,
It's a structured grid but the z-spacing varies. So I don't think PiecewiseMultilinear helps unfortunately. Although I suppose I could perhaps interpolate it onto a regularly spaced grid.

**Reply by GiudGiud** [reply]
Hello

As long as it is a structured grid you can work directly with PiecewiseMultiLinear. See this folder
https://github.com/idaholab/moose/tree/next/test/tests/functions/piecewise_multilinear
for examples. `twoD1.txt` has irregular spacing in X and Y 

The AXIS Z parameter wont have regular spacing but this is ok

Guillaume

**Reply by hsheldon** [reply]
Thanks Guillaume,
The z spacing varies in x and y so that won't work unfortunately. I think I will have to resample onto a regular grid.

**Reply by GiudGiud** [reply]
ah ok. Does it vary often? If it only varies like 2 or 3 times you could define 3 different functions then take the sum

I also have a PR up for general CSV interpolation (nearest neighbor), you could use that. No guarantees on when it ll be merged.
https://github.com/idaholab/moose/pull/19118
If you do want that, let me know I ll fix the merge conflicts

**Reply by hsheldon** [reply]
It varies everywhere, it's not predictable at all. Anyway, I wrote a script to interpolate to a regular grid, then PiecewiseMultilinear worked perfectly - very happy with the result.
